src/resmom/linux/alps.c

```diff
@@ -6397,129 +6665,916 @@ alps_inventory(void)
+				attr = "resources_available.PBScrayseg";
+				if (vn_addvnr(vnlp, vname, attr, "0", ATR_TYPE_STR, atype, NULL) == -1)
+					goto bad_vnl;
+			}
+			attr = "resources_available.ncpus";
+			ncpus_per_knl = atoi(node_group->compute_units);
+			snprintf(utilBuffer_knl, sizeof(utilBuffer_knl), "%d", ncpus_per_knl);
+			if (vn_addvnr(vnlp, vname, attr, utilBuffer_knl, 0, 0, NULL) == -1)
+				goto bad_vnl;
+
+			/* avlmem is conventional DRAM mem. avlmem = page_size_kb * page_count. */
+			attr = "resources_available.mem";
+			snprintf(utilBuffer_knl, sizeof(utilBuffer_knl), "%skb", node_group->avlmem);
+			if (vn_addvnr(vnlp, vname, attr, utilBuffer_knl, 0, 0, NULL) == -1)
+				goto bad_vnl;
+
+			attr = "current_aoe";
+			snprintf(utilBuffer_knl, sizeof(utilBuffer_knl), "%s_%s",
+				node_group->numa_cfg, node_group->hbm_cfg);
+			if (vn_addvnr(vnlp, vname, attr, utilBuffer_knl, 0, 0, NULL) == -1)
+				goto bad_vnl;
+
+			/* hbmem is high bandwidth mem (MCDRAM) in megabytes. */
+			attr = "resources_available.hbmem";
+			snprintf(utilBuffer_knl, sizeof(utilBuffer_knl), "%smb", node_group->hbmsize);
+			if (vn_addvnr(vnlp, vname, attr, utilBuffer_knl, 0, 0, NULL) == -1)
+				goto bad_vnl;
+		}
+		/* We have no further use for this array of KNL node ids. */
+		if (nid_arr) {
+			free(nid_arr);
+			nid_arr = NULL;
+		}
+	}
+
+	return;
+
+bad_vnl:
+	snprintf(log_buffer, sizeof(log_buffer), "Creation of Cray KNL vnodes failed with name %s", vname);
+	log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_NODE, LOG_DEBUG, __func__, log_buffer);
+	/*
+	 * Don't free nv since it might be important in the dump.
+	 */
+	abort();
+}
+
+/**
+ *
+ * @brief Check if this Node Group needs to be considered for KNL processing.
+ *        We are only interested in KNL Nodes that have "role" set to "batch" and
+ *	  "state" set to "up".
+ *	  The attributes "numa_cfg", "hbmsize", "hbm_cfg" not being empty ("") implies
+ *	  they pertain to KNL Nodes.
+ *
+ * @param[in] ptrNodeGrp Pointer to the current Node Group in the System XML Response.
+ *
+ * @return int
+ * @retval 1 Indicates that the Node Group should not be considered.
+ * @retval 0 Indicates that the Node Group should be considered.
+ *
+ */
+static int
+exclude_from_KNL_processing(basil_system_element_t *ptrNodeGrp)
+{
+	if ((strcmp(ptrNodeGrp->role, BASIL_VAL_BATCH_SYS) != 0) ||
+	    (strcmp(ptrNodeGrp->state, BASIL_VAL_UP_SYS) != 0) ||
+	    ((strcmp(ptrNodeGrp->numa_cfg, "") == 0) &&
+	     (strcmp(ptrNodeGrp->hbmsize, "") == 0) &&
+	     (strcmp(ptrNodeGrp->hbm_cfg, "") == 0)))
+		return 1;
+	else
+		return 0;
+}
+
+/**
+ *
+ * @brief KNL Nodes are specified in 'Rangelist' format in a string e.g. "12,13,14-18,21".
+ * 	  Extract Node IDs from this string and store them in an integer array.
+ *
+ * @param[in] nidlist String containing rangelist of Nodes.
+ * @param[out] ptr_count Total number of Nodes in this list.
+ *
+ * @return int *
+ * @retval This is an integer array containing Node IDs.
```

Comment:
In the node structure (basil_node_t) node_id is defined to be a long.  I understand that you are converting from a string, however instead of converting into an int I think you should make it of type long to match the rest of the vnode creation.